pipeline/src/pipeline.rs

```diff
@@ -708,7 +711,12 @@ impl<T: FieldElement> Pipeline<T> {
     // but give it different witnesses and generate different proofs.
     // The previous alternative to this was cloning the entire pipeline.
     pub fn rollback_from_witness(&mut self) {
-        self.artifact.witness_and_publics = None;
+        if let Some(old_arc) = &self.artifact.witness_and_publics {
+            // Clone the public part
+            let public = old_arc.1.clone();
+            // Replace with an empty witness and the preserved public
+            self.artifact.witness_and_publics = Some(Arc::new((Vec::new(), public)));
```

Comment:
Why do we keep the publics?

Reply:
I think this causes a few other changes in the file. In my mind, scalar publics could almost be considered part of the witness (you solve for them, you need them to evaluate constraints, ...), so I don't see why they are treated differently.